File: python/2022/19.py
from timeit import default_timer as timer
from dataclasses import dataclass
from pulp import LpMaximize, LpProblem, LpVariable, PULP_CBC_CMD, LpStatus
from util.util import getRunsFromIniNewlineSep
import pandas as pd
import os
import re


@dataclass
class Blueprint:
  # Eact dict is resourceName -> resourceAmount. E.g.,
  # oreRobot = {'ore': 4, 'clay': 18}
  oreRobot: dict[str, int]
  clayRobot: dict[str, int]
  obsidianRobot: dict[str, int]
  geodeRobot: dict[str, int]

  def __str__(self):
    details = ''
    details += 'oreRobot = ' + ' + '.join([f'{v}*{k}'for k,v in
      self.oreRobot.items()]) +'\n'
    details += 'clayRobot = ' + ' + '.join([f'{v}*{k}'for k,v in
      self.clayRobot.items()]) +'\n'
    details += 'obsidianRobot = ' + ' + '.join([f'{v}*{k}'for k,v in
      self.obsidianRobot.items()]) +'\n'
    details += 'geodeRobot = ' + ' + '.join([f'{v}*{k}'for k,v in
      self.geodeRobot.items()]) +'\n'
    return details

  def __repr__(self):
    return self.__str__()

@dataclass
class Simulation:
  name: str
  blueprints: list[Blueprint]

  def __str__(self):
    details = ''
    details += f'Name       : {self.name}\n'
    details += (
        "blueprints :\n"
        + "\n".join([blueprint.__str__() for blueprint in self.blueprints])
        + "\n"
    )
    return details

  def __repr__(self):
    return self.__str__()

def inputfile() -> str:
    return os.path.dirname(os.path.realpath(__file__)) + '/19.txt'

def getSimulations() -> list[Simulation]:
  sims = getRunsFromIniNewlineSep(inputfile())
  result: list[Simulation] = []
  for [name, contents] in sims:
    blueprints: list[Blueprint] = []

    for line in contents:
      data: dict[str, dict[str, int]] = {}
      matches = re.findall(r'Each ([a-z]+) robot costs (.+?)\.', line)
      assert matches

      for robot, costs in matches:
        data[robot] = {}
        for cost in costs.split(' and '):
          count, resource = cost.split(' ')
          data[robot][resource] = int(count)

      blueprints.append(Blueprint(data['ore'], data['clay'], data['obsidian'],
        data['geode']))
    result.append(Simulation(name, blueprints))
  return result

def solve(blueprint: Blueprint, time = 24) -> float:
  T = time
  model = LpProblem(sense=LpMaximize)

  # How many available resources you have at time 0<=t<=24
  ores = [
    LpVariable(name=f"ores[{i}]", lowBound=0, cat="Integer") for i in range(T + 1)
  ]
  clays = [
    LpVariable(name=f'clays[{i}]', lowBound=0, cat='Integer') for i in range(T + 1)
  ]
  obsidians = [
    LpVariable(name=f'obsidians[{i}]', lowBound=0, cat='Integer') for i in range(T + 1)
  ]
  geodes = [
    LpVariable(name=f'geodes[{i}]', lowBound=0, cat='Integer') for i in range(T + 1)
  ]

  # How many robots you have at time 0<=t<=24
  ore_robots = [
    LpVariable(name=f"ore_robots[{i}]", lowBound=0, cat="Integer")
    for i in range(T + 1)
  ]
  clay_robots = [
    LpVariable(name=f"clay_robots[{i}]", lowBound=0, cat="Integer")
    for i in range(T + 1)
  ]
  obsidian_robots = [
    LpVariable(name=f"obsidian_robots[{i}]", lowBound=0, cat="Integer")
    for i in range(T + 1)
  ]
  geode_robots = [
    LpVariable(name=f"geode_robots[{i}]", lowBound=0, cat="Integer")
    for i in range(T + 1)
  ]

  # Boolean variables. Whether a robot was built or not at time 0<=t<=24
  ore_robots_built = [
    LpVariable(name=f"ore_robots_built[{i}]", cat="Binary") for i in range(T + 1)
  ]
  clay_robots_built = [
    LpVariable(name=f"clay_robots_built[{i}]", cat="Binary") for i in range(T + 1)
  ]
  obsidian_robots_built = [
    LpVariable(name=f"obsidian_robots_built[{i}]", cat="Binary")
      for i in range(T + 1)
  ]
  geode_robots_built = [
    LpVariable(name=f"geode_robots_built[{i}]", cat="Binary") for i in range(T + 1)
  ]

  ####### CONSTRAINTS #######
  # Initially, you have nothing at time = 0 but one ore robot.
  model += (ores[0] == 0)
  model += (clays[0] == 0)
  model += (obsidians[0] == 0)
  model += (geodes[0] == 0)

  model += (ore_robots[0] == 1)
  model += (clay_robots[0] == 0)
  model += (obsidian_robots[0] == 0)
  model += (geode_robots[0] == 0)

  model += (ore_robots_built[0] == 0)
  model += (clay_robots_built[0] == 0)
  model += (obsidian_robots_built[0] == 0)
  model += (geode_robots_built[0] == 0)

  # Ores at time t is equal to ores at t-1 plus # of ore robots at t-1
  # MINUS all the ores used at time t.
  for t in range(1, T + 1):
    model += (
      ores[t]
      == ores[t - 1]
      + ore_robots[t - 1]
      - blueprint.oreRobot.get("ore", 0) * ore_robots_built[t]
      - blueprint.clayRobot.get("ore", 0) * clay_robots_built[t]
      - blueprint.obsidianRobot.get("ore", 0) * obsidian_robots_built[t]
      - blueprint.geodeRobot.get("ore", 0) * geode_robots_built[t]
    )
    model += (
      clays[t]
      == clays[t - 1]
      + clay_robots[t - 1]
      - blueprint.oreRobot.get("clay", 0) * ore_robots_built[t]
      - blueprint.clayRobot.get("clay", 0) * clay_robots_built[t]
      - blueprint.obsidianRobot.get("clay", 0) * obsidian_robots_built[t]
      - blueprint.geodeRobot.get("clay", 0) * geode_robots_built[t]
    )
    model += (
      obsidians[t]
      == obsidians[t - 1]
      + obsidian_robots[t - 1]
      - blueprint.oreRobot.get("obsidian", 0) * ore_robots_built[t]
      - blueprint.clayRobot.get("obsidian", 0) * clay_robots_built[t]
      - blueprint.obsidianRobot.get("obsidian", 0) * obsidian_robots_built[t]
      - blueprint.geodeRobot.get("obsidian", 0) * geode_robots_built[t]
    )
    model += (geodes[t] == geodes[t - 1] + geode_robots[t - 1])

  # Ores built at time t is equal to ores built at t-1 + robots built at t-1.
  for t in range(1, T + 1):
    model += (ore_robots[t] == ore_robots[t - 1] + ore_robots_built[t - 1])
    model += (clay_robots[t] == clay_robots[t - 1] + clay_robots_built[t - 1])
    model += (obsidian_robots[t] == obsidian_robots[t - 1] +
        obsidian_robots_built[t - 1])
    model += (geode_robots[t] == geode_robots[t - 1] + geode_robots_built[t - 1])

  # One robot can be built per time t
  for t in range(T + 1):
    model += (ore_robots_built[t] + clay_robots_built[t] +
        obsidian_robots_built[t] + geode_robots_built[t] <= 1)

  obj_func = geodes[T] # Maximize the number of geodes at the end of 24mins
  model += obj_func

  start = timer()
  model.solve(PULP_CBC_CMD(msg=False))
  end = timer()
  # printModel(model, end - start)
  return model.objective.value()

def printModel(model: LpProblem, duration) -> None:
  print(f"status: {model.status}, {LpStatus[model.status]}")
  print(f"Elapsed time (ms): {(duration) * 1000}\n\n")
  allvars = model.variables()
  allvars.sort(key=lambda v:int(re.findall(r'\d+', v.name)[0]))
  df = pd.DataFrame(
    map(
      lambda v: [
        # the names are of the form ores_robot_10_
        re.findall(r"\d+", v.name)[0], # this will be 10
        re.findall(r"^(.+)_\d", v.name)[0], # will be ores_robot
        v.value(),
      ],
      allvars,
    ),
    # the index to these rows will be just the row number, and there are 24 *
    # model vars number of rows.
    columns=["time", "name", "value"],
  # pivot will convert a table with columns time, name, and value
  # where time is [1, 24] and names are ore/clay/obsidian/geode.
  # and the rows are combinations of those. We can pivot similar to
  # Sheets where the first column are unique values of time [1,24]
  # and the other columns are the unique names for the robots and the
  # values are model variables value.
  ).pivot(index="time", columns="name", values="value")
  # print(df) just prints some of the rows. df.to_string() will print all
  print(df.to_string())
  df.to_csv('day19.csv')

def run(sim: Simulation) -> float:
  sum = 0
  for idx, blueprint in enumerate(sim.blueprints):
    obj = solve(blueprint)
    sum += (idx + 1) * obj
    print(idx, obj)
  return sum
sims = getSimulations()

def run2(sim: Simulation) -> float:
  mult = 1
  for idx, blueprint in enumerate(sim.blueprints[:3]):
    obj = solve(blueprint, 32)
    mult *= obj
    print(idx, obj)
  return mult

sims = getSimulations()
print(run(sims[1]))

File: src/2022/19.ts
import { assert } from '~/util/assert';
import { timer } from '~/util/Timer';
import { getRunsFromIniNewlineSep } from '~/util/util';
import input from './19.txt';

interface Cost {
  ore: number;
  clay: number;
  obsidian: number;
}

interface Blueprint {
  oreRobot: Cost;
  clayRobot: Cost;
  obsidianRobot: Cost;
  geodeRobot: Cost;
}

interface Simulation {
  name: string;
  blueprints: Blueprint[];
}

type DropRobot<T> = {
  [K in keyof T as K extends `${infer I}Robot` ? I : K]: T[K];
};

function getSimulations(): Simulation[] {
  return getRunsFromIniNewlineSep(input).map((sim) => ({
    name: sim.name,
    blueprints: sim.content.map((line) => {
      const matches = assert(
        line.matchAll(/Each ([a-z]+) robot costs (.+?)\./g)
      );
      const blueprint: Partial<Blueprint> = {};
      for (const [, robotName, costsStr] of matches) {
        const costObj: Cost = { ore: 0, clay: 0, obsidian: 0 };
        for (const cost of costsStr.split(' and ')) {
          const [amount, costName] = cost.split(' ');
          costObj[costName as keyof Cost] = Number(amount);
        }
        blueprint[(robotName + 'Robot') as keyof Blueprint] = costObj;
      }
      return blueprint as Blueprint;
    }),
  }));
}

interface State {
  resourcesAvailable: {
    ore: number;
    clay: number;
    obsidian: number;
    geode: number;
  };
  robots: {
    ore: number;
    clay: number;
    obsidian: number;
    geode: number;
  };
}

function serialize(state: State, timeRemaining: number) {
  const { resourcesAvailable: ra, robots: r } = state;
  return (
    `${ra.ore},${ra.clay},${ra.obsidian},${ra.geode}` +
    `${r.ore},${r.clay},${r.obsidian},${r.geode},${timeRemaining}`
  );
}

export function solve1(blueprint: Blueprint) {
  let gmaxGeodes = 0; // global max geodes
  // serialized state to # of geodes
  const visited = new Map<string, number>();
  const helper = (state: State, timeRemaining: number): number => {
    if (timeRemaining <= 0) return 0;

    const stateStr = serialize(state, timeRemaining);
    if (visited.has(stateStr)) return visited.get(stateStr)!;
    // Heuristic #1
    if (potentialGeodes({ ...state, timeRemaining }) < gmaxGeodes) {
      return 0;
    }

    let newState: State | undefined = collectResources(state);
    let maxGeodes = newState.resourcesAvailable.geode;
    // Option #1: don't build anything. Just have the robots collect resources.
    // Heuristic #2
    if (addMoreResources(newState, blueprint)) {
      maxGeodes = Math.max(maxGeodes, helper(newState, timeRemaining - 1));
      gmaxGeodes = Math.max(gmaxGeodes, maxGeodes);
    }
    if ((newState = buildRobot(blueprint, 'geode', state))) {
      // Option#2: See if you can build a geode robot
      maxGeodes = Math.max(maxGeodes, helper(newState, timeRemaining - 1));
      gmaxGeodes = Math.max(gmaxGeodes, maxGeodes);
    }
    if ((newState = buildRobot(blueprint, 'obsidian', state))) {
      // Option#3: See if you can build a obsidian robot
      maxGeodes = Math.max(maxGeodes, helper(newState, timeRemaining - 1));
      gmaxGeodes = Math.max(gmaxGeodes, maxGeodes);
    }
    if ((newState = buildRobot(blueprint, 'clay', state))) {
      // Option#4: See if you can build a clay robot
      maxGeodes = Math.max(maxGeodes, helper(newState, timeRemaining - 1));
      gmaxGeodes = Math.max(gmaxGeodes, maxGeodes);
    }
    if ((newState = buildRobot(blueprint, 'ore', state))) {
      // Option#5: See if you can build an ore robot
      maxGeodes = Math.max(maxGeodes, helper(newState, timeRemaining - 1));
      gmaxGeodes = Math.max(gmaxGeodes, maxGeodes);
    }
    visited.set(stateStr, maxGeodes);
    return maxGeodes;
  };
  const ans = helper(
    {
      resourcesAvailable: { ore: 0, clay: 0, obsidian: 0, geode: 0 },
      robots: { ore: 1, clay: 0, obsidian: 0, geode: 0 },
    },
    24
  );
  return ans;
}

/**
 * If you have enough resources to build any robot, don't bother with collecting
 * more resources (causing this function to return false). You should instead be
 * building bots + collecting resources at the same time.
 */
function addMoreResources(state: State, blueprint: Blueprint): boolean {
  const { ore, clay, obsidian } = state.resourcesAvailable;
  const { oreRobot, clayRobot, obsidianRobot, geodeRobot } = blueprint;
  if (
    ore <
    Math.max(oreRobot.ore, clayRobot.ore, obsidianRobot.ore, geodeRobot.ore)
  ) {
    return true;
  } else if (
    clay <
    Math.max(oreRobot.clay, clayRobot.clay, obsidianRobot.clay, geodeRobot.clay)
  ) {
    return true;
  } else if (
    obsidian <
    Math.max(
      oreRobot.obsidian,
      clayRobot.obsidian,
      obsidianRobot.obsidian,
      geodeRobot.obsidian
    )
  ) {
    return true;
  }
  return false;
}

/**
 * Given there are N minutes left, what's the max number of geodes can you make
 * if you have infinite resources? It's the sum of
 *    * number of geodes you already have
 *    * number of geode robots you have * timeRemaining
 *        - E.g., if you have two robots and 5 remaining minutes, you'll make 10 geodes.
 *    * (timeRemaining-1) + (timeRemaining-2) + .. + 1
 *             = (timeRemaining) * (timeRemaining - 1) / 2
 *        - This is assuming you have infinite resources and you can build a
 *        geode every minute. If you have 2 miutes left, you can build a robot
 *        at t=2, and that robot can create 1 geode.
 *        - If you have 3 minutes left, you build a robot#1 at t=3, which would
 *        collect 2 geodes before time ran out. You build robot #2 at t=2, which
 *        would collect 1 geode.
 *        - With 4 minute left, you'd end up with 3 + 2 + 1 = 6 geode.
 *        - With N minute left, you'd end up with (N-1) + (N-2) + .. + 1 =
 *          N(N-1)/2 geodes.
 */
function potentialGeodes({
  timeRemaining,
  ...state
}: State & { timeRemaining: number }) {
  const futurePotential =
    timeRemaining <= 1 ? 0 : (timeRemaining * (timeRemaining - 1)) / 2;
  return (
    state.resourcesAvailable.geode +
    state.robots.geode * timeRemaining +
    futurePotential
  );
}

function collectResources<T extends State>(state: T): T {
  const newState = {
    ...state,
    resourcesAvailable: { ...state.resourcesAvailable },
    robots: { ...state.robots },
  };
  newState.resourcesAvailable.ore += newState.robots.ore;
  newState.resourcesAvailable.clay += newState.robots.clay;
  newState.resourcesAvailable.obsidian += newState.robots.obsidian;
  newState.resourcesAvailable.geode += newState.robots.geode;
  return newState;
}

/**
 * @return The updated state after building the specified robot and collecting
 * resources. If the robot cannot be built due to lack of resources, it returns
 * undefined.
 */
function buildRobot<T extends State>(
  blueprint: Blueprint,
  type: keyof DropRobot<Blueprint>,
  state: T
): T | undefined {
  const cost = blueprint[(type + 'Robot') as keyof Blueprint];
  const {
    resourcesAvailable: { ore, clay, obsidian },
  } = state;
  if (ore >= cost.ore && clay >= cost.clay && obsidian >= cost.obsidian) {
    const updatedState = {
      ...state,
      resourcesAvailable: {
        ...state.resourcesAvailable,
        ore: state.resourcesAvailable.ore - cost.ore + state.robots.ore,
        clay: state.resourcesAvailable.clay - cost.clay + state.robots.clay,
        obsidian:
          state.resourcesAvailable.obsidian -
          cost.obsidian +
          state.robots.obsidian,
        geode: state.resourcesAvailable.geode + state.robots.geode,
      },
      robots: { ...state.robots, [type]: state.robots[type] + 1 },
    };
    return updatedState;
  }
  return undefined;
}

export function run() {
  const sims = getSimulations().slice(1);
  for (const sim of sims) {
    for (const [idx, blueprint] of sim.blueprints.entries()) {
      timer.run(() => {
        console.log(solve1(blueprint));
      }, `${sim.name} - ${idx + 1}`);
    }
  }
}
